### DISCUSSION TITLE
Fetch always with localStorage

### DISCUSSION THREAD

**Discussion by sofimiazzi** [pmndrs/jotai#1962] [Q&A]
Every time I store and retrieve information in localStorage using atomWithStorage, the atom starts with the initial state. However, in the initial state, I make an external API call. I would like this call to be made only once, and when using Jotai with localStorage, this call occurs every time I reload my components. What would be the best approach to solve this problem?

```ts
const fetchLanguage = async () => {
  const { data } = await geolocationApi.get("client-info");
  return data.userLanguages[0];
};

export const settingsAtom = atomWithStorage("settings", {
  theme: "dark",
  language: await fetchLanguage(),
  isMuted: true,
  playbackRate: 1,
});

```

**Comment by dai-shi** [comment]
> the atom starts with the initial state.

We are trying to fix this in #1958. Stay tuned.

> However, in the initial state, I make an external API call.

Your code doesn't work as you call it outside atoms. You need two atoms.

```js
// This will be possible when #1958 lands.
const internalSettingsAtom = atomWithStorage(...);
const settingsAtom = atom(async (get) => get(baseAtom) || {
  theme: "dark",
  language: await fetchLanguage(),
  isMuted: true,
  playbackRate: 1,
}));
```
(it's still suboptimal, but that's the basic idea.)
